### ISSUE TITLE
[feat] AbortController support for `invoke` Promises

### ISSUE THREAD

**Issue by douweschulte** [tauri-apps/tauri#8351]
### Describe the problem

I was looking into a way to cancel/abort the async Rust code I am running in tauri. My use case is that I have operations which could take up to a couple of minutes, and if the user realises they set some parameter wrong they would like to cancel to start the operation again. But the exact use case does not necessarily matter, there are many reasons to want to abort a called function.

### Describe the solution you'd like

I was reading up on the `AbortController`/`AbortSignla` API that is implemented for many basal JS promises (like fetch amongst others). I tried to use this with the build in Promise from tauri `invoke` but this did not respond to the abort request. Slimmed down my code looks like to following:
```js
// Some basic setup
const controller = new AbortController();

// The time intensive operation, started by pressing a button in the interface
async function annotate_spectrum() {
    let promise = invode("annotate_spectrum"), {..arguments});
    promise.then((result) => /* on success */).catch((result)=>/*on error*/);
    promise.signal = controller.signal;
}

// A separate function to abort the running operation, for now started by using a different button
function abort() {
  controller.abort("User manual abort");
}
```
This above code runs, the `AbortController` responds correctly to the abort() command, and even the promise itself stores the abortion signal correctly in its state. 

My feature suggestion is for the promise to abort in response to this `AbortSignal`. 

### Additional context

It could very well be that this is already possible if I rewrite my current code, if so I am sorry for raising the issue, but would greatly appreciate if someone could point me in the correct direction. Also I would be happy to help if any substantial dev work is needed. I am quite well versed in Rust, but my JS skills are a bit lacking.

While searching for solutions I found this [question from another user with AbortController in the answer](https://flapr.io/topic/details/abort-invoke-promises-in-tauri/25175) (this question was cross posted to stackoverflow).

**Comment by amrbashir** [comment]
TBH I think this could be an awesome addition to Tauri and can be implemented for async commands (not sure about sync commands though), don't you think @lucasfernog @JonasKruckenberg?

> While searching for solutions I found this [question from another user with AbortController in the answer](https://flapr.io/topic/details/abort-invoke-promises-in-tauri/25175) (this question was cross posted to stackoverflow).

All this solution does, is that it aborts the promise in JS but doesn't cancel the work in Rust-side.

Anyways, this can be implemented in your own code if you wish, basically you will need to have:
- A store (maybe `HashMap` or `BTreeMap`) where you store all your Rust futures
- 3 commands:
  - one that initiates the async logic and stores it in the HashMap and returns an identifier that could be used by the other two commands
  - one that cancels the future which basically just drops the future from the store
  - one that awaits the future and sends the result, if the cancel was called, it will return nothing, since the future has been dropped, otherwise return the future result
- then your JS function encapsulates all of this:
  ```js
  async function doSomething({signal}) {
  	const id = await invoke("start-do-something");
  	const abortHandler = () => invoke("cancel-do-something", { id });
  	signal?.addEventListener("abort", abortHandler);
  	const actualResult = await invoke("finish-do-something", { id });
  	signal?.removeEventListener("abort", abortHandler);
  	return actualResult;
  }
  ```

An example of this can be found in the `http` plugin here https://github.com/tauri-apps/plugins-workspace/blob/387c2f9e0ce4c75c07ffa3fd76391a25b58f5daf/plugins/http

**Comment by douweschulte** [comment]
Thanks a lot for the insight! And thanks a lot for the potential way of building this. I had been thinking about something like this but it was way less solid then your suggestion here. I was indeed only thinking about async, cancelling sync commands is incongruent with the concept of sync commands.

**Comment by JonasKruckenberg** [comment]
Yeah I think this would  be a cool feature, albeit one with more caveats and footguns than could be listed (future cancellation haha) but I'm unsure how easy it'd be to implement in our current setup? Seems like we would need a way for the frontend to pass along something like a requestID that can be used to identify tasks 🤔

**Comment by amrbashir** [comment]
Indeed, we would basically generalize what we did for the http plugin to all invoke calls and with the newly added `resources_table` in `dev` branch, this got a lot easier IMO.

**Comment by douweschulte** [comment]
Is there any interest for this to be done within Tauri? Otherwise @amrbashir could you point me to that example you described, then I can try my hand at implementing it on my side.

**Comment by amrbashir** [comment]
@douweschulte This is still desired but haven't had the time to implement just yet.

Here is how `@tauri-apps/plugin-http` implements fetch:
- Rust: https://github.com/tauri-apps/plugins-workspace/blob/406e6f484cdc13d35c50fb949f7489ca9eeccc44/plugins/http/src/commands.rs
- Javascript: https://github.com/tauri-apps/plugins-workspace/blob/406e6f484cdc13d35c50fb949f7489ca9eeccc44/plugins/http/guest-js/index.ts

The idea is that you have split your command into 3 commands, let's take `fetch` as an example:

1. `fetch` (a prepare command), basically creates two async future, one that makes an HTTP request, and one that will be used as a cancel token. The command will then store them both in `Manager::resources_table` without awaiting any of them, and instead returns a `RresourceId` to be used by other two commands.
2. `fetch_cancel`, using the `ResourceId`, this command sends a cancel request to the second future.
3. `fetch_send`, using the `ResourceId`, this command awaits the two stored future, and if the cancel request future completes first, then we drop the HTTP request (i.e cancelling it) and return an error to JS, otherwise we complete the HTTP request and return it to JS.

**Comment by douweschulte** [comment]
Thanks so much for the additional information! I will try my hand at an implementation on my side, but will keep my eyes peeled for an implementation inside Tauri.
